xls/simulation/cocotb/cocotb_xls.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2023 The XLS Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cocotb
+
+from cocotb.triggers import RisingEdge
+from cocotb.binary import BinaryValue
+from cocotb.handle import SimHandleBase
+
+from cocotb_bus.drivers import BusDriver
+from cocotb_bus.monitors import BusMonitor
+
+from typing import Any, List
+
+
+class XLSChannelDriver(BusDriver):
+    _signals = ["data", "rdy", "vld"]
+
+    def __init__(
+        self, entity: SimHandleBase, name: str, clock: SimHandleBase, **kwargs: Any
+    ):
+        BusDriver.__init__(self, entity, name, clock, **kwargs)
+
+        self.bus.data.setimmediatevalue(0)
+        self.bus.vld.setimmediatevalue(0)
+
+    async def write(self, data: List[BinaryValue], sync: bool = True) -> None:
+        if sync:
+            await RisingEdge(self.clock)
+
+        for word in data:
+            self.bus.vld.value = 1
+            self.bus.data.value = word
+
+            while True:
+                await RisingEdge(self.clock)
+                if self.bus.rdy.value:
```

Comment:
shouldn't we also wait for `rdy` before transmitting the first bit?

Reply:
I used the same convention as in AXI Stream interface. The fragment of the text below comes from the second chapter of the [AMBA AXI-Stream Protocol Specification](https://developer.arm.com/documentation/ihi0051/latest/):

> * The TVALID and TREADY handshake determines when information is passed across the interface. A two-way flow control mechanism enables both the Transmitter and Receiver to control the rate at which the data and control information is transmitted across the interface.
> * For a transfer to occur, both TVALID and TREADY must be asserted. Either TVALID or TREADY can be asserted first, or both can be asserted in the same ACLK cycle.
> * **A Transmitter is not permitted to wait until TREADY is asserted before asserting TVALID. Once TVALID is asserted, it must remain asserted until the handshake occurs.**
> * A Receiver is permitted to wait for TVALID to be asserted before asserting TREADY. It is permitted that a Receiver asserts and deasserts TREADY without TVALID being asserted

It is possible to allow exactly opposite behavior, so to allow the `Transmitter` to wait on `TREADY`, but then the `Reciever` should not be permitted to wait on `TVALID` to guarantee that there will be no deadlocks in the system.
Since the AXI Stream protocol is popular and existing IP-cores use this protocol, I decided to be consent with it.

Reply:
> Since the AXI Stream protocol is popular and existing IP-cores use this protocol, I decided to be in line with it.

aren't we trying to map to the XLS channel semantic here? (since this is for `XLSChannelDriver`), we should be able to verify in the generated verilog for a receiver proc what is the expected behavior of the drive wrt to the `rdy` and `vld` assertions.

Reply:
looking at the documentation there seems to be multiple options available, see https://google.github.io/xls/codegen_options/#io-behavior

Reply:
I changed the code to use the default behavior of XLS